test/e2e/policies.test-suite/topology-aware/c4pmem4/test05-guarantee-memory/code.var.sh

```diff
@@ -1,40 +1,152 @@
-CRI_RESMGR_OUTPUT="cat cri-resmgr.output.txt"
+CRI_RESMGR_OUTPUT="cat cri-resmgr.output.txt | tr -d '\0'"
+CRI_RESMGR_ROTATE="echo > cri-resmgr.output.txt"
 
-# pod0: 2 containers, both get their own socket and take > 50 % of their socket's mem
-CPU=200m MEM=5G CONTCOUNT=2 create guaranteed
+podno=0
+kubectl delete pod --all --now
+
+# account for being done with test for the current pod
+nextpod () {
+    podno=$((podno+1))
+}
+
+# print current pod name
+pod () {
+    echo pod$podno
+}
+
+# print current container name, by default for current pod
+container () {
+    local _p _c
+    case $# in
+        0) _p=${podno}; _c=0;;
+        1) _p=${podno}; _c=$1;;
+        2) _p=$1; _c=$2;;
+        *)
+            echo podWTFcHUHwhy${#}ARGS
+            return 1
+            ;;
+    esac
+    case $_p in
+        +*|-*) _p=$((${podno}$_p));;
+    esac
+    echo pod${_p}c${_c}
+}
+
+# rotate cri-resmgr logs
+rotate_log () {
+    vm-command "$CRI_RESMGR_ROTATE"
+}
+
+###########################################################################
+# test #1: squeeze multiple containers in every NUMA node
+#
+# We squeeze an increasing number of containers in all NUMA node pools
+# in a loop. For every iteration we calculate the usable amount of CPU
+# and memory based on the available number of NUMA nodes and the amount
+# of CPU and memory per NUMA node. We use a conservative estimate for
+# the amount of memory available per NUMA node because some of them will
+# have a sizeable allocation by the kernel.
+#
+
+rotate_log
+
+# use conservative estimate for available memory per node
+PER_NODE_MEM=$((1500+4000))
+PER_NODE_CPU=1000
+PER_NODE_PMEM=1
+NODE_COUNT_TOTAL=4
+# All nodes have only a single CPU. Thus, with any (< 1000m) CPU reservation
+# we'll have one node (#0) fully reserved for kube-system containers. Hence,
+# our (usable) node count is one less than the total one.
+NODE_COUNT=$((NODE_COUNT_TOTAL - 1))
+
+for pernode in 2 3 4; do
+    cpu=$(echo "scale=3;0.75*$PER_NODE_CPU/$pernode" | bc | cut -d '.' -f1)
+    mem=$(echo "scale=3;0.75*$PER_NODE_MEM/$pernode" | bc | cut -d '.' -f1)
+    CPU=${cpu}m MEM=${mem}Mi CONTCOUNT=$((pernode*NODE_COUNT)) create guaranteed
+
+    echo "Verify that any pod's containers were not raised to guarantee memory"
+    echo ""
+    vm-command "$CRI_RESMGR_OUTPUT | grep upward" && {
+        pp mems
+        error "Unexpected memset upward expansion detected!"
+    }
+
+    echo "Verify that all containers are pinned to a single NUMA node"
+    echo ""
+    c=0; while [ "$c" -lt "$pernode" ]; do
+        verify "len(mems['$(container $c)']) == $((1+PER_NODE_PMEM))"
```

Comment:
We have pernode*NODE_COUNT containers. Does this somehow check all of them? 🤔

Reply:
No. Fixed.